core-spi-test/src/main/java/org/ehcache/internal/store/StoreEventListenerTest.java

```diff
@@ -0,0 +1,333 @@
+    assertThat(lastValue, is(nullValue()));
+    assertThat(listener.expired, hasItem(k));
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SPITest
+  public void testRemoveOnExpiration() throws Exception {
+    TestTimeSource timeSource = new TestTimeSource();
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        null, null, null, Expirations.timeToLiveExpiration(new Duration(1, TimeUnit.MILLISECONDS))), timeSource);
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    timeSource.advanceTime(1);
+    assertThat(kvStore.remove(k, v), is(false));
+    assertThat(listener.expired, hasItem(k));
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SPITest
+  public void testReplaceTwoArgsOnExpiration() throws Exception {
+    TestTimeSource timeSource = new TestTimeSource();
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        null, null, null, Expirations.timeToLiveExpiration(new Duration(1, TimeUnit.MILLISECONDS))), timeSource);
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    timeSource.advanceTime(1);
+    assertThat(kvStore.replace(k, v), is(nullValue()));
+    assertThat(listener.expired, hasItem(k));
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SPITest
+  public void testReplaceThreeArgsOnExpiration() throws Exception {
+    TestTimeSource timeSource = new TestTimeSource();
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        null, null, null, Expirations.timeToLiveExpiration(new Duration(1, TimeUnit.MILLISECONDS))), timeSource);
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    timeSource.advanceTime(1);
+    assertThat(kvStore.replace(k, v, v2), is(false));
+    assertThat(listener.expired, hasItem(k));
+  }
+
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  @SPITest
+  public void testComputeOnExpiration() throws Exception {
+    TestTimeSource timeSource = new TestTimeSource();
+    Store<K, V> kvStore = factory.newStore(factory.newConfiguration(factory.getKeyType(), factory.getValueType(),
+        null, null, null, Expirations.timeToLiveExpiration(new Duration(1, TimeUnit.MILLISECONDS))), timeSource);
+    TestStoreEventListener listener = new TestStoreEventListener();
+    kvStore.enableStoreEventNotifications(listener);
+    kvStore.put(k, v);
+    timeSource.advanceTime(1);
+    kvStore.compute(k, new BiFunction<K, V, V>() {
+      @Override
+      public V apply(K mappedKey, V mappedValue) {
+        return v2;
+      }
+    });
+    assertThat(kvStore.get(k).value(), is(v2));
```

Comment:
This shouldn't be here right?  The compute itself should trigger the expiration and listener fire.